poncho/src/poncho/library_network_code.py

```diff
@@ -265,48 +321,82 @@ def main():
 
         # check if library cores and function slots are valid
         if args.function_slots > args.library_cores:
-            print(
-                "Library code: function slots cannot be more than library cores",
-                file=sys.stderr,
-            )
+            stdout_timed_message("error: function slots cannot be more than library cores")
             exit(1)
         elif args.function_slots < 1:
-            print("Library code: function slots cannot be less than 1", file=sys.stderr)
+            stdout_timed_message("error: function slots cannot be less than 1")
             exit(1)
         elif args.library_cores < 1:
-            print("Library code: library cores cannot be less than 1", file=sys.stderr)
+            stdout_timed_message("error: library cores cannot be less than 1")
             exit(1)
-        thread_limit = args.library_cores // args.function_slots
+
+        try:
+            thread_limit = args.library_cores // args.function_slots
+        except Exception as e:
+            stdout_timed_message(f"error: {e}")
+            exit(1)
+
+        # check if the in_pipe_fd and out_pipe_fd are valid
+        try:
+            fcntl.fcntl(args.in_pipe_fd, fcntl.F_GETFD)
+            fcntl.fcntl(args.out_pipe_fd, fcntl.F_GETFD)
+        except IOError as e:
+            stdout_timed_message(f"error: pipe fd closed\n{e}")
+            exit(1)
+
+        stdout_timed_message(f"library task starts running in process {os.getpid()}")
+        stdout_timed_message(f"hostname             {socket.gethostname()}")
+        stdout_timed_message(f"task id              {args.task_id}")
+        stdout_timed_message(f"worker pid           {args.worker_pid}")
+        stdout_timed_message(f"library pid          {os.getpid()}")
+        stdout_timed_message(f"input fd             {args.in_pipe_fd}")
+        stdout_timed_message(f"output fd            {args.out_pipe_fd}")
+        stdout_timed_message(f"library cores        {args.library_cores}")
+        stdout_timed_message(f"function slots       {args.function_slots}")
+        stdout_timed_message(f"thread limit         {thread_limit}")
+        stdout_timed_message(f"functions installed\n{get_installed_function_code()}")
```

Comment:
Again, the names of the functions, or the entire function code?

Reply:
@dthain It logs the entire function code, for example:
```
06/05/24 12:37:28.008669 functions installed
def my_func(size=100):
    import numpy
    import pandas
    import sklearn.preprocessing
    import sklearn.decomposition
    import scipy.stats
    import scipy.special
    import matplotlib.pyplot
    import hashlib
    import os
    print('hello')
    matrix_a = numpy.random.rand(int(size / 2), int(size / 2))
    matrix_b = numpy.random.rand(int(size / 2), int(size / 2))
    data = numpy.matmul(matrix_a, matrix_b)
    data = numpy.random.randn(size, 2) * 200 + numpy.random.rand(size, 2) * 100
    df = pandas.DataFrame(data, columns=['A', 'B'])
    df_normalized = df.apply(scipy.stats.zscore)
    scaler_standard = sklearn.preprocessing.StandardScaler().fit_transform(df_normalized)
    scaler_minmax = sklearn.preprocessing.MinMaxScaler().fit_transform(df_normalized)
    pca = sklearn.decomposition.PCA(n_components=2)
    pca_result = pca.fit_transform(scaler_standard)
    transformed_data = scipy.special.expit(pca_result)
    matplotlib.pyplot.figure(figsize=(12, 6))
    matplotlib.pyplot.subplot(1, 1, 1)
    matplotlib.pyplot.scatter(pca_result[:, 0], pca_result[:, 1], c='blue', label='PCA Result')
    matplotlib.pyplot.title('PCA Plot')
    unique_string = str(transformed_data.tolist()) + str(pca_result.tolist()) + str(scaler_minmax.tolist()) + str(df_normalized.values.tolist())
    hash_object = hashlib.sha256(unique_string.encode())
    hash_string = hash_object.hexdigest()
    return hash_string
def estimate_pi(num_samples=30000000):
    import random
    points_inside_circle = 0
    for _ in range(num_samples):
        x, y = (random.uniform(-1, 1), random.uniform(-1, 1))
        distance_to_center = x ** 2 + y ** 2
        if distance_to_center <= 1:
            points_inside_circle += 1
    pi_estimate = 4 * points_inside_circle / num_samples
    return pi_estimate
```